src/Symfony/Component/ObjectMapper/ObjectMapper.php

```diff
@@ -176,6 +178,58 @@ private function getSourceValue(object $source, object $target, mixed $value, \S
             $value = $this->applyTransforms($mapping, $value, $source, $target);
         }
 
+        if ($mapping instanceof MapCollection) {
+            if (!\is_iterable($value)) {
+                return $value;
+            }
+
+            $mappedCollection = [];
+            foreach ($value as $item) {
+                if (\is_object($item)) {
+                    $mappedCollection[] = $this->map($item, $mapping->target);
+                } else {
+                    $mappedCollection[] = $item;
+                }
+            }
+
+            return $mappedCollection;
+        }
+
+        if ($mapping instanceof MapTree) {
+            if (!\is_iterable($value)) {
+                return $value;
+            }
+
+            $mappedTree = [];
+            foreach ($value as $node) {
+                if (!\is_object($node)) {
+                    $mappedTree[] = $node;
+                    continue;
+                }
+
+                $mappedNode = $this->map($node, $mapping->target);
+
+                $children = $this->getRawValue($node, $mapping->childrenProperty);
+                $mappedChildren = [];
+
+                if (\is_iterable($children)) {
+                    foreach ($children as $child) {
+                        $mappedChildren[] = $this->map($child, $mapping->target);
+                    }
+                }
+
+                if ($this->propertyAccessor) {
+                    $this->propertyAccessor->setValue($mappedNode, $mapping->childrenProperty, $mappedChildren);
+                } else {
+                    $mappedNode->{$mapping->childrenProperty} = $mappedChildren;
+                }
+
+                $mappedTree[] = $mappedNode;
+            }
+
+            return $mappedTree;
```

Comment:
I don't think its a good idea to include these functionalities directly in the ObjectMapper, we should instead create new kinds of Mappers that compose with the ObjectMapper